thread/common/threadhelpers.cpp

```diff
@@ -50,52 +50,57 @@ intptr_t
 omrthread_spinlock_acquire(omrthread_t self, omrthread_monitor_t monitor)
 {
 	volatile uintptr_t *target = (volatile uintptr_t *)&monitor->spinlockState;
-	intptr_t result = 0;
+	intptr_t result = -1;
 	uintptr_t oldState = J9THREAD_MONITOR_SPINLOCK_UNOWNED;
 	uintptr_t newState = J9THREAD_MONITOR_SPINLOCK_OWNED;
+
+	omrthread_library_t lib = self->library;
+
 #if defined(OMR_THR_JLM)
-	J9ThreadMonitorTracing *tracing = (self->library->flags & J9THREAD_LIB_FLAG_JLM_ENABLED) ? monitor->tracing : NULL;
+	J9ThreadMonitorTracing *tracing = NULL;
+	if (J9_ARE_ALL_BITS_SET(lib->flags, J9THREAD_LIB_FLAG_JLM_ENABLED)) {
+		tracing = monitor->tracing;
+	}
 #endif /* OMR_THR_JLM */
 
-#if defined(OMR_THR_SPIN_WAKE_CONTROL)
-	BOOLEAN spinning = FALSE;
+	uintptr_t spinCount3Init = monitor->spinCount3;
+	uintptr_t spinCount2Init = monitor->spinCount2;
+	uintptr_t spinCount1Init = monitor->spinCount1;
 
- 	if (monitor->spinThreads >= self->library->maxSpinThreads) {
- 		result = -1;
- 		goto done;
- 	}
+	uintptr_t spinCount3 = spinCount3Init;
+	uintptr_t spinCount2 = spinCount2Init;
 
-	VM_AtomicSupport::add(&monitor->spinThreads, 1);
-	spinning = TRUE;
+#if defined(OMR_THR_SPIN_WAKE_CONTROL)
+	BOOLEAN spinning = TRUE;
+ 	if ((monitor->spinThreads < lib->maxSpinThreads) && J9_ARE_NO_BITS_SET(monitor->flags, J9THREAD_MONITOR_DISABLE_SPINNING)) {
+ 		VM_AtomicSupport::add(&monitor->spinThreads, 1);
+ 	} else {
+ 		spinning = FALSE;
+ 		spinCount3Init = 1;
+ 		spinCount2Init = 1;
+ 		spinCount1Init = 1;
+ 		spinCount3 = spinCount3Init;
+ 		spinCount2 = spinCount2Init;
```

Comment:
Are these 2 lines correct?  The are going to set spinCount3 and spinCount2 to 1 because you just modified the values of the Init variables.  If they are correct it may be more clear to just set them to 1 explicitly.